File: examples/sidecars/node/sidecar.mjs
#!/usr/bin/env node
// Minimal RAIC Overlay sidecar in Node.js (ESM).
//
// Implements docs/plugins/sidecar-protocol.md:
//   raic.init       -> responds with { name, version, methods }
//   ping            -> responds with { reply: 'pong', receivedAt }
//   echo            -> mirrors params back
//   raic.shutdown   -> exits cleanly within 5 seconds
//
// Build a single-file executable with e.g. `bun build --compile`, `pkg`,
// or Node's `--experimental-sea-config` to ship as `bin/<platform>/sidecar.exe`.

import { stdin, stdout, stderr, exit, env } from 'node:process';
import { createInterface } from 'node:readline';

const PLUGIN_ID = env.RAIC_PLUGIN_ID ?? '?';
stderr.write(`node sidecar starting for plugin ${PLUGIN_ID}\n`);

function write(obj) {
  stdout.write(JSON.stringify(obj) + '\n');
}

function nowRfc3339() {
  return new Date().toISOString();
}

const rl = createInterface({ input: stdin, crlfDelay: Infinity });

for await (const line of rl) {
  let msg;
  try {
    msg = JSON.parse(line);
  } catch (e) {
    stderr.write(`invalid JSON: ${e.message}\n`);
    continue;
  }

  const { method, id, params } = msg;

  switch (method) {
    case 'raic.init':
      write({
        jsonrpc: '2.0',
        id,
        result: {
          name: 'sidecar-node',
          version: '0.1.0',
          methods: ['ping', 'echo'],
        },
      });
      break;

    case 'ping':
      write({
        jsonrpc: '2.0',
        id,
        result: { reply: 'pong', receivedAt: nowRfc3339() },
      });
      break;

    case 'echo':
      write({ jsonrpc: '2.0', id, result: params });
      break;

    case 'raic.shutdown':
      stderr.write('shutdown received; exiting\n');
      exit(0);
      break;

    default:
      write({
        jsonrpc: '2.0',
        id,
        error: { code: -32601, message: `method '${method}' not found` },
      });
  }
}

File: examples/sidecars/python/sidecar.py
#!/usr/bin/env python3
"""
Minimal RAIC Overlay sidecar in Python.

Implements docs/plugins/sidecar-protocol.md:
  - raic.init        → responds with { name, version, methods }
  - ping             → responds with { reply: 'pong', receivedAt }
  - echo             → mirrors params back
  - raic.shutdown    → exits cleanly within 5 seconds

Usage from a plugin manifest:

    "sidecar": {
      "platforms": {
        "windows-x86_64": { "bin": "bin/windows-x86_64/sidecar.exe" }
      }
    }

(Wrap with PyInstaller or similar to ship as a single exe.)
"""
from __future__ import annotations

import datetime
import json
import os
import sys


def now_rfc3339() -> str:
    return datetime.datetime.now(datetime.timezone.utc).isoformat()


def write(obj: dict) -> None:
    sys.stdout.write(json.dumps(obj) + "\n")
    sys.stdout.flush()


def log(msg: str) -> None:
    # Host captures stderr lines and tags them with the plugin id.
    sys.stderr.write(msg + "\n")
    sys.stderr.flush()


def main() -> int:
    plugin_id = os.environ.get("RAIC_PLUGIN_ID", "?")
    log(f"python sidecar starting for plugin {plugin_id}")

    for line in sys.stdin:
        try:
            msg = json.loads(line)
        except json.JSONDecodeError as e:
            log(f"invalid JSON: {e}")
            continue

        method = msg.get("method")
        msg_id = msg.get("id")

        if method == "raic.init":
            write(
                {
                    "jsonrpc": "2.0",
                    "id": msg_id,
                    "result": {
                        "name": "sidecar-python",
                        "version": "0.1.0",
                        "methods": ["ping", "echo"],
                    },
                }
            )
        elif method == "ping":
            write(
                {
                    "jsonrpc": "2.0",
                    "id": msg_id,
                    "result": {"reply": "pong", "receivedAt": now_rfc3339()},
                }
            )
        elif method == "echo":
            write(
                {
                    "jsonrpc": "2.0",
                    "id": msg_id,
                    "result": msg.get("params"),
                }
            )
        elif method == "raic.shutdown":
            log("shutdown received; exiting")
            return 0
        else:
            write(
                {
                    "jsonrpc": "2.0",
                    "id": msg_id,
                    "error": {
                        "code": -32601,
                        "message": f"method {method!r} not found",
                    },
                }
            )

    return 0


if __name__ == "__main__":
    sys.exit(main())
